tests/support/client.rs

```diff
@@ -8,43 +8,57 @@ use self::tokio::{
     io::{AsyncRead, AsyncWrite},
     net::TcpStream,
 };
+use bytes::{Buf, BufMut};
+use rustls::{ClientConfig, ClientSession};
+use std::io::{Read, Write};
+use std::sync::Arc;
 use support::bytes::IntoBuf;
 use support::hyper::body::Payload;
+use tokio_rustls::TlsStream;
+use webpki::DNSName;
 
+type ClientError = hyper::Error;
 type Request = http::Request<Bytes>;
 type Response = http::Response<BytesBody>;
-type Sender = mpsc::UnboundedSender<(Request, oneshot::Sender<Result<Response, String>>)>;
+type Sender = mpsc::UnboundedSender<(Request, oneshot::Sender<Result<Response, ClientError>>)>;
+type TlsConfig = (Arc<ClientConfig>, DNSName);
 
 #[derive(Debug)]
 pub struct BytesBody(hyper::Body);
 
-pub fn new<T: Into<String>>(addr: SocketAddr, auth: T) -> Client {
-    http2(addr, auth.into())
+pub fn new<T: Into<String>>(addr: SocketAddr, auth: T, tls: Option<TlsConfig>) -> Client {
+    http2(addr, auth.into(), tls)
 }
 
-pub fn http1<T: Into<String>>(addr: SocketAddr, auth: T) -> Client {
+pub fn http1<T: Into<String>>(addr: SocketAddr, auth: T, tls: Option<TlsConfig>) -> Client {
```

Comment:
Let's consider making a new function that takes a `tls` parameter instead of an `Option` like we are doing here. It will make this change significantly smaller since we will not need to update all the calls to this function in the rest of the tests.

I'm think something like `pub fn tls_http1<T: Into<String>>(..., tls_config: TlsConfig) -> Client {}`.

The current `http1`/`http2` functions can just pass `None` into `new` so that we can continue making one `Client`.

Reply:
+1 for @kleimkuhler's comment --- I'd really prefer to not have to pass `None` into all test client constructors in the future.